langchain/schema.py

```diff
@@ -390,3 +390,9 @@ class OutputParserException(Exception):
     """
 
     pass
+
+
+class BaseDocumentProcessor(ABC):
+    @abstractmethod
+    def process(self, documents: List[Document]) -> List[Document]:
```

Comment:
It may be nice to separate the single doc and the batch doc interfaces

```python

class BaseDocumentProcessor(ABC):
    @abstractmethod
    def process_document(self, documents: Document) -> Document:
        """Process documents."""

    def process_documents(self, documents: List[Document]) -> List[Document]:
      return [self.process_document(doc) for doc in documents]
```

This would provide a default for loop implementation while allowing inherting classes to over-ride if batching can be made more efficient. 

Batch composition is easy: 

```python
class  SequentialBatchProcessor():  # <-- Only supports batch case, not streaming
   def __init__(self, processors) -> None:
      self.processors = []

    def process_documents(self, documents: List[Document]) -> List[Document]:
      for processor in self.processors:
       docs = processor(docs)
       return docs
```

It would nice to figure out if it's possible to make a streaming version (more complicated given that a splitter can be used as a processor with interface of doc -> [doc] or [doc] -> [doc])

Reply:
@hwchase17 let's also add an `async` version for batch processing, we can provide a default implementation